### PR TITLE
unit testing fixes and pyFileSystem 2 support

### PR DESCRIPTION
**Pull request by tcpan**
Added support for pyFileSystem 2 in keyrings.alt  as my project requires both pyFS 2 and keyrings.alt.

Also fixed keyrings.alt unit tests, a lot of which were not run because they did not subclass unittest.TestCase nor followed pytest class naming conventions.

In my fork, I also created a branch (py27_pyfs2) off tags/v3.2.0 with similar changes but for python 2.7.   This is to support my users who may be staying with python 2.

### CODE REVIEW

**Comment by mitya57** [comment]
@jaraco Ping!

File: keyrings/alt/pyfs.py [hunk 1]
```diff
@@ -9,14 +9,44 @@
...
+import fs.errors
+import fs.path
+
+try:
+    if FS_VER < version.parse("2.0.0"):
+        import fs.remote
+    else:
+        import fs.wrap
 except ImportError:
-    pass
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:19]: Previously, the behavior was not to raise an error when fs wasn't available. This new approach causes an ImportError when fs isn't available. That means that any users without fs installed will get an error when importing keyring. For example, I applied this diff.

```diff
diff --git a/keyrings/alt/pyfs.py b/keyrings/alt/pyfs.py
index 08addd2..cab6713 100644
--- a/keyrings/alt/pyfs.py
+++ b/keyrings/alt/pyfs.py
@@ -17,6 +17,7 @@ try:
     import fs.remote
 except ImportError:
     pass
+import doesnotexist
 
 
 def has_pyfs():
```

Then:

```
keyrings.alt master $ .tox/python/bin/python -c "import keyring"
Error initializing plugin EntryPoint(name='pyfs', value='keyrings.alt.pyfs', group='keyring.backends').
Traceback (most recent call last):
  File "/Users/jaraco/code/main/keyrings.alt/.tox/python/lib/python3.8/site-packages/keyring/backend.py", line 198, in _load_plugins
    init_func = ep.load()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/jaraco/code/main/keyrings.alt/keyrings/alt/pyfs.py", line 20, in <module>
    import doesnotexist
ModuleNotFoundError: No module named 'doesnotexist'
```

These modules are designed to be lenient to errors and to signal their viability or not through the 'priority' method.

File: keyrings/alt/pyfs.py [hunk 2]
```diff
@@ -9,14 +9,44 @@
 from keyring.backend import KeyringBackend, NullCrypter
 from . import keyczar
 
+# support getting module version
+import pkg_resources
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:13]: pkg_resources is discouraged for this use. Let's use importlib.metadata instead. Regardless, you need to declare the dependency.

File: keyrings/alt/pyfs.py [hunk 3]
```diff
@@ -9,14 +9,44 @@
...
+
+# support parsing version string
+from packaging import version
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:16]: Need to declare the dependency on packaging.

File: keyrings/alt/tests/backend.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
+# coding: utf-8
```
Comment (jaraco) [inline keyrings/alt/tests/backend.py:1]: Copying this file from the other project is a bad idea. It means this code will evolve independently and need to have fixes applied to both. Moreover, I can't tell from this commit whether this code was copied exactly as it appears in keyring.testing or if other changes were applied before the commit. Let's find a way to do this without copy/paste code.

Comment (jaraco) [reply keyrings/alt/tests/backend.py:1]: Oh, I see - much of the duplicated functionality is removed in later commits.

File: keyrings/alt/pyfs.py [hunk 5]
```diff
@@ -9,14 +9,44 @@
...
+    FS_VER = version.parse(pkg_resources.get_distribution("fs").version)
+except pkg_resources.DistributionNotFound:
+    raise ImportError
+
+import fs.osfs
+import fs.errors
+import fs.path
+
+try:
+    if FS_VER < version.parse("2.0.0"):
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:29]: I'd like to see this fs compatibility behavior extracted into a module or function.

File: keyrings/alt/pyfs.py [hunk 6]
```diff
@@ -47,24 +75,72 @@ class BasicKeyring(KeyringBackend):
     def __init__(self, crypter, filename=None, can_create=True, cache_timeout=None):
         super(BasicKeyring, self).__init__()
         self._crypter = crypter
-        def_fn = os.path.join(platform_.data_root(), self.__class__._filename)
-        self._filename = filename or def_fn
+        if filename is None:
+            self._filename = os.path.join(
+                platform_.data_root(), self.__class__._filename
+            )
+        elif filename.endswith("/"):
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:82]: This code seems to be changing the meaning of the filename parameter from being a filename to being a filename_or_directory, but only treating it like a directory if it ends in a posix-like slash. I don't care for that interface and I'm not sure what problem this solves. Probably this change should have its own issue and separate change.

File: keyrings/alt/pyfs.py [hunk 7]
```diff
@@ -261,6 +284,9 @@ def __init__(self, filename=None, can_create=True, cache_timeout=None):
             cache_timeout=cache_timeout,
         )
 
+    def __del__(self):
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:287]: I'm pretty sure this code is unneeded.

File: keyrings/alt/pyfs.py [hunk 8]
```diff
@@ -25,9 +55,7 @@ def has_pyfs():
     Should return False even when Mercurial's Demand Import allowed import of
     fs.*.
     """
-    with errors.ExceptionRaisedContext() as exc:
-        fs.opener.opener
-    return not bool(exc)
+    return FS_VER is not None
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:58]: The docstring no longer matches the implementation.

File: keyrings/alt/pyfs.py [hunk 9]
```diff
@@ -253,6 +274,8 @@ class PlaintextKeyring(BasicKeyring):
     """Unencrypted Pyfilesystem Keyring
     """
 
+    _filename = 'uncrypted_pyf_pass.cfg'
```
Comment (jaraco) [inline keyrings/alt/pyfs.py:277]: Looks like this user-facing incompatible change got snuck in.

**Comment by jaraco** [comment]
In #38, I've started working on the commits that fix the problems with the test suite not running due to the migration from unittest to pytest.

**Comment by jaraco** [comment]
I've rebased the two pyFileSystem changes on top of the master following #38. That makes it a little easier to focus in this PR on those changes.

**Review by jaraco — CHANGES_REQUESTED** [review]
Thanks for the PR! I've made a few comments above. If we can address these, I'd be delighted to adopt the changes.

File: keyrings/alt/tests/test_pyfs.py [hunk 10]
```diff
@@ -69,20 +98,35 @@ def test_handles_preexisting_keyring(self):
...
+            e = sys.exc_info()[1]
+            if e.errno != errno.ENOENT:  # No such file or directory
+                raise
 
 
 class TestUnencryptedLocalPyfilesystemKeyringSubDir(PyFSBackend):
     """Test using local temp files with no encryption"""
 
-    keyring_dir = os.path.join(tempfile.mkdtemp(), 'more', 'sub', 'dirs')
+    keyring_root = tempfile.mkdtemp()
```
Comment (jaraco) [inline keyrings/alt/tests/test_pyfs.py:115]: This directory gets created unconditionally but is only cleaned up if the tests run. Better to put this creation behavior in the _cleanup_for_fs fixture before `yield` (and rename it to match new behavior).

File: keyrings/alt/tests/test_pyfs.py [hunk 11]
```diff
@@ -100,14 +144,41 @@ def init_keyring(self):
...
     """Test using local temp files with encryption"""
 
+    keyring_root = tempfile.mkdtemp()
+    keyring_filename = '%s/keyring.cfg' % keyring_root
+
     def init_keyring(self):
-        return pyfs.EncryptedKeyring(ReverseCrypter(), filename='temp://keyring.cfg')
+        return pyfs.EncryptedKeyring(ReverseCrypter(), filename=self.keyring_filename)
+
+    @pytest.fixture(autouse=True, scope="class")
```
Comment (jaraco) [inline keyrings/alt/tests/test_pyfs.py:153]: This looks like copy-pasta. Please extract into a shared fixture.

File: keyrings/alt/tests/test_pyfs.py [hunk 12]
```diff
@@ -11,6 +14,18 @@
 from keyring.testing.backend import BackendBasicTests
 from keyring.testing.util import random_string
 
+# support getting module version
+import pkg_resources
```
Comment (jaraco) [inline keyrings/alt/tests/test_pyfs.py:18]: This code looks like a repeat of other code. Probably just use the code from the module.

**Comment by jaraco** [comment]
Since the comments have lingered for a long time, I'm closing this languishing request, but I'm happy to revisit should interest revive.

**Pull request closed without merging**